### PR TITLE
feat: LU copy & paste & delete with unified CRUD lib

### LINKED ISSUE #2273: Need to support LU intent copy / delete / move
- select one node or several nodes (inputs with user utterance)
- copy and paste in the dialog, no lu intent copied 
- move to the new dialog, no lu content copied

### LINKED ISSUE #2406: [dialog factory] create lg template when an lg field has a default value
## Describe the bug

Currently, when adding actions that include default values for lg, like QnA step, the editors complain that there are no templates in the lg file. To fix this, the form should create one if there is no value and a default is provided.

## To Reproduce

1. Create a new QNA maker step in a brand new bot without any lg (like EmptyBot).

## Expected behavior

No error should be shown.

## Screenshots

image

### LINKED ISSUE #2726: LG template name not correct
## Describe the bug
refs #2597

1. LG in SendActivity
image
2. LG after copied
image

They should follow
image

## Browser

- [ ] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Edge

## OS

- [ ] macOS
- [ ] Windows
- [ ] Ubuntu

### PR DESCRIPTION
**Pull request by yeze322**
## Description
1. Fix LU related copy / paste / delete behaviors ('Move' function requires #2742 be merged).
2. Migrate to a unified CRUD lib to bring unified *Action / LU / LG* resources' *Create / Read / Update / Delete / Copy / Paste / Move* experience.

close #2273 
close #2406 
close #2726 
refs #2699 
refs #2312 (only applied to Visual Editor)

Changes:
1. Implemented several shared hooks as unified CRUD lib
2. Copy / Paste behavior will always invoke action constructor (#2406 )
3. Broke 'Move' operation into several atomic CRUD action
4. make Visual Editor consumes the shared CRUD lib

#### External Resource API
- `useLgApi` -> operate lgTemplate
- `useLuApi` -> operate luIntent
#### Adaptive Dialog API
- `useActionApi` -> construct / copy / destruct actions
- ~~(not included) `useTriggerApi`~~
- `useDialogApi` -> operate dialog file
#### Editing API
- `useDialogEditApi` -> copySelection; cutSelection; pasteSelection; deleteSelection; moveSelection

### CODE REVIEW

File: Composer/packages/extensions/extension/src/hooks/useDialogApi.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+export const useDialogApi = () => {
+  const {
+    shellApi: { getDialog, saveDialog, createDialog },
+  } = useShellApi();
+
+  async function createDialogCompleted(): Promise<string | undefined> {
+    const newDialogId = await createDialog([]);
+    // TODO: @lei9444 be able to new dialog data here.
+    return newDialogId;
+  }
```
Comment (yeze322) [inline Composer/packages/extensions/extension/src/hooks/useDialogApi.ts:15]: @lei9444 could you help to implement the `createDialogComplteted` method here?
It requires the newly created dialog data is readable via `shellApi.getDialog`

Comment (lei9444) [reply Composer/packages/extensions/extension/src/hooks/useDialogApi.ts:15]: @yeze322 fix it in #2724

File: Composer/packages/client/src/shell/useShell.ts [hunk 2]
```diff
@@ -125,32 +105,44 @@ export function useShell(source: EventSource): { api: ShellApi; data: ShellData
     }
   }, [schemas, projectId]);
 
+  dialogMapRef.current = dialogsMap;
+
   const api: ShellApi = {
+    getDialog: (dialogId: string) => {
+      return dialogMapRef.current[dialogId];
```
Comment (yeze322) [inline Composer/packages/client/src/shell/useShell.ts:112]: @lei9444 there is a bug here related to the usage of `ref.current`
If I moved some actions to a new dialog, composer data will be updated but the actual file won't be updated until next time I edit the file.

Comment (lei9444) [reply Composer/packages/client/src/shell/useShell.ts:112]: @yeze322 I have fix this bug in #2742 , It's not related to the ref. It's a bug in file persistence sync logic

**Comment by cwhitten** [comment]
@yeze322 when we Move a selected series of actions with .lu data into a new dialog we also should set the recognizer field on the new dialog to be the same as the recognizer field in current dialog.

File: Composer/packages/client/src/shell/useShell.ts (hunk 2 above)
Comment (yeze322) [reply Composer/packages/client/src/shell/useShell.ts:112]: refs #2742

**Comment by yeze322** [comment]
@cwhitten I got it. the `recognizer` field in new dialog will be set to `<dialogName>.lu`

**Comment by cwhitten** [comment]
@yeze322 LG templates are not copied over appropriately when the using "Move"

**Comment by yeze322** [comment]
> @yeze322 LG templates are not copied over appropriately when the using "Move"

yes, with only this PR, new lg/lu/dialog file will not be updated correctly.
it depends on Leilei's PR #2742

File: Composer/packages/extensions/extension/src/hooks/useDialogApi.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+export const useDialogApi = () => {
+  const {
+    shellApi: { getDialog, saveDialog, createDialog },
+  } = useShellApi();
+
+  async function createDialogCompleted(): Promise<string | null> {
+    const newDialogId = await createDialog([]);
+    // TODO: @lei9444 be able to new dialog data here.
+    return newDialogId;
```
Comment (yeze322) [inline Composer/packages/extensions/extension/src/hooks/useDialogApi.ts:14]: remove the TODO here

File: Composer/packages/extensions/extension/src/hooks/useDialogApi.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  }
+
+  return {
+    createDialog: createDialogCompleted,
```
Comment (yeze322) [inline Composer/packages/extensions/extension/src/hooks/useDialogApi.ts:18]: function name

File: Composer/packages/extensions/visual-designer/src/editors/ObiEditor.tsx [hunk 5]
```diff
@@ -144,108 +86,90 @@ export const ObiEditor: FC<ObiEditorProps> = ({
...
+          const newDialogWithActions = await insertActions(
+            newDialogId,
+            newDialogData,
+            `${'triggers'}[0].${'actions'}`,
+            0,
+            actionsToBeMoved
+          );
+          if (actionsContainLuIntent(actionsToBeMoved)) {
+            // auto assign recognizer type to lu
+            newDialogWithActions.recognizer = `${newDialogId}.lu`;
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/editors/ObiEditor.tsx:146]: recognizer logic leak. Consider adding a `updateRecognizer()` method in `dialogEditApi`

File: Composer/packages/lib/shared/src/types/shell.ts [hunk 6]
```diff
@@ -60,7 +62,10 @@ export interface ShellApi {
   updateLgTemplate: (id: string, templateName: string, templateStr: string) => Promise<void>;
   removeLgTemplate: (id: string, templateName: string) => Promise<void>;
   removeLgTemplates: (id: string, templateNames: string[]) => Promise<void>;
-  updateLuIntent: (id: string, intentName: string, intent: LuIntentSection | null) => void;
+  getLuIntent: (id: string, intentName: string) => LuIntentSection | undefined;
+  getLuIntents: (id: string) => LuIntentSection[];
+  addLuIntent: (id: string, intentName: string, intent: LuIntentSection | undefined) => Promise<void>;
+  updateLuIntent: (id: string, intentName: string, intent: LuIntentSection | undefined) => Promise<void>;
   updateRegExIntent: (id: string, intentName: string, pattern: string) => void;
   removeLuIntent: (id: string, intentName: string) => void;
```
Comment (yeze322) [inline Composer/packages/lib/shared/src/types/shell.ts:70]: group lu Apis together

**Pull request merged**